File: replay-tests/multi-module-app/app/test/ui/hello/BaseSpec.java
package ui.hello;

import com.codeborne.selenide.Configuration;
import hello.HelloWorldApp;
import java.util.concurrent.atomic.AtomicBoolean;
import org.junit.jupiter.api.BeforeEach;
import play.Play;

public class BaseSpec {
  private static final AtomicBoolean appStarted = new AtomicBoolean(false);

  @BeforeEach
  public void setUp() {
    System.setProperty("webdriver.http.factory", "jdk-http-client");
    if (appStarted.get()) return;
    startApp();
  }

  private static synchronized void startApp() {
    if (appStarted.get()) return;

    int port = new HelloWorldApp().start("test");

    Configuration.baseUrl = "http://localhost:" + port;
    Play.configuration.setProperty("application.baseUrl", Configuration.baseUrl);

    appStarted.set(true);
  }
}

File: replay-tests/helloworld-kotlin/test/ui/hello/BaseSpec.kt
package ui.hello

import com.codeborne.selenide.Configuration
import org.junit.Before
import org.openqa.selenium.net.PortProber.findFreePort
import play.Play
import play.server.Server

open class BaseSpec {
  private val play = Play()

  @Before
  fun setUp() {
    val playStarter = Thread({
      play.init("test")
      play.start()
      val port = findFreePort()
      Server(play, port).start()
      Configuration.baseUrl = "http://localhost:$port"
      Play.configuration.setProperty("application.baseUrl", Configuration.baseUrl)
    }, "Play! starter thread")
    playStarter.start()
    playStarter.join()
    Configuration.browser = "chrome"
    Configuration.headless = true
  }
}